test/001_highsnr_stdio_audio/test_callback_rx.py

```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Wed Dec 23 07:04:24 2020
+
+@author: DJ2LS
+"""
+
+import ctypes
+from ctypes import *
+import pathlib
+import pyaudio
+import sys
+import logging
+import time
+import threading
+import sys
+import argparse
+import numpy as np
+sys.path.insert(0,'..')
+import codec2
+
+#--------------------------------------------GET PARAMETER INPUTS  
+parser = argparse.ArgumentParser(description='FreeDATA audio test')
+parser.add_argument('--bursts', dest="N_BURSTS", default=1, type=int)
+parser.add_argument('--framesperburst', dest="N_FRAMES_PER_BURST", default=1, type=int)
+parser.add_argument('--mode', dest="FREEDV_MODE", type=str, choices=['datac0', 'datac1', 'datac3'])
+parser.add_argument('--audiodev', dest="AUDIO_INPUT_DEVICE", default=-1, type=int,
+                    help="audio device number to use, use -2 to automatically select a loopback device")  
+parser.add_argument('--debug', dest="DEBUGGING_MODE", action="store_true")  
+parser.add_argument('--timeout', dest="TIMEOUT", default=10, type=int, help="Timeout (seconds) before test ends")  
+parser.add_argument('--list', dest="LIST", action="store_true", help="list audio devices by number and exit")  
+
+args = parser.parse_args()
+
+if args.LIST:
+    p = pyaudio.PyAudio()
+    for dev in range(0,p.get_device_count()):
+        print("audiodev: ", dev, p.get_device_info_by_index(dev)["name"])
+    quit()
+   
+N_BURSTS = args.N_BURSTS
+N_FRAMES_PER_BURST = args.N_FRAMES_PER_BURST
+AUDIO_INPUT_DEVICE = args.AUDIO_INPUT_DEVICE
+MODE = codec2.FREEDV_MODE[args.FREEDV_MODE].value
+DEBUGGING_MODE = args.DEBUGGING_MODE
+TIMEOUT = args.TIMEOUT
+
+# AUDIO PARAMETERS
+AUDIO_FRAMES_PER_BUFFER = 2400*2  # <- consider increasing if you get nread_exceptions > 0
+MODEM_SAMPLE_RATE = codec2.api.FREEDV_FS_8000
+AUDIO_SAMPLE_RATE_RX = 48000
+
+# make sure our resampler will work
+assert (AUDIO_SAMPLE_RATE_RX / MODEM_SAMPLE_RATE) == codec2.api.FDMDV_OS_48
+
+
+# ------------------------------------------------ PYAUDIO CALLBACK
+def callback(data_in48k, frame_count, time_info, status):
+    x = np.frombuffer(data_in48k, dtype=np.int16)
+    x.tofile(frx)    
+    x = resampler.resample48_to_8(x)
+    audio_buffer.push(x)    
```

Comment:
I don't think audio buffer is thread safe,  e.g. the callback and main thread could be making audio_buffer calls at the same time.  So we might need some syncronisation objects like mutexes (of the Python equivalent) around the audio buffer push and pop functions.

The [codec2_fifo](https://github.com/drowe67/codec2/blob/master/src/codec2_fifo.h) design is thread safe by design, but we'd need to re-write it in Python, or call the C version from Python.

Reply:
I'm surprised again about the hidden features of codec2 :) However, as I wrote above, I'm not sure if we really need the callback mode. Maybe its enough running the entire audio part within a own thread. 

This is also an area I need to learn more to get a better understanding.

Reply:
From reading a few stack exchange threads, looking at Horus GUI, and my experience with PortAudio on freedv-gui .. I get the feeling that the callback model might be best.